core/pta/ocall.c

```diff
@@ -0,0 +1,344 @@
+
+		switch (ca_pt) {
+		case TEE_PARAM_TYPE_NONE:
+		case TEE_PARAM_TYPE_VALUE_INPUT:
+			break;
+		case TEE_PARAM_TYPE_VALUE_INOUT:
+		case TEE_PARAM_TYPE_VALUE_OUTPUT:
+			ca_param->value.a = rpc_params[n].u.value.a;
+			ca_param->value.b = rpc_params[n].u.value.b;
+			break;
+		case TEE_PARAM_TYPE_MEMREF_INPUT:
+			if (rpc_params[n].u.memref.size !=
+			    ca_param->memref.size)
+				return TEE_ERROR_BAD_PARAMETERS;
+			if (ADD_OVERFLOW(mobj_offs, rpc_params[n].u.memref.size,
+					 &mobj_offs))
+				return TEE_ERROR_SECURITY;
+			break;
+		case TEE_PARAM_TYPE_MEMREF_INOUT:
+		case TEE_PARAM_TYPE_MEMREF_OUTPUT:
+			if (rpc_params[n].u.memref.size > ca_param->memref.size)
+				return TEE_ERROR_BAD_PARAMETERS;
+			memcpy(ca_param->memref.buffer,
+			       PTR_ADD(mobj_va, mobj_offs),
+			       rpc_params[n].u.memref.size);
+			ca_param->memref.size = rpc_params[n].u.memref.size;
+			if (ADD_OVERFLOW(mobj_offs, rpc_params[n].u.memref.size,
+					 &mobj_offs))
+				return TEE_ERROR_SECURITY;
+			break;
+		default:
+			return TEE_ERROR_BAD_PARAMETERS;
+		}
+	}
+
+	return TEE_SUCCESS;
+}
+
+/*
+ * TA Interface:
+ * -------------
+ * params[0]: VALUE IN/OUT
+ *   IN: < HOST CMD ID,  Param Types   >
+ *  OUT: < CMD Ret Val, CMD Ret Origin >
+ * params[1]: < [INOUT] TEE_Param[] >
+ * params[2]: < - >
+ * params[3]: < - >
+ *
+ * RPC Interface:
+ * --------------
+ * params[0]: VALUE IN/OUT
+ *    IN: < HOST CMD ID,      -     ,        -       >
+ *   OUT: <  Unchanged , CMD Ret Val, CMD Ret Origin >
+ * params[1]: < [IN] UUID Octets Hi, [IN] UUID Octets Lo, - >
+ * params[2..5]: < OCALL Params, if any >
+ *
+ * There are two return codes:
+ *   res: Did sending the RPC work?
+ *   ca_ret: If sending the RPC worked, what is the return value of the OCALL?
+ */
+static TEE_Result ocall_send(struct tee_ta_session *s, uint32_t param_types,
+			     TEE_Param params[TEE_NUM_PARAMS])
+{
+	const uint32_t pt1 = TEE_PARAM_TYPE_GET(param_types, 0);
+	const uint32_t pt2 = TEE_PARAM_TYPE_GET(param_types, 1);
+	const uint32_t pt3 = TEE_PARAM_TYPE_GET(param_types, 2);
+	const uint32_t pt4 = TEE_PARAM_TYPE_GET(param_types, 3);
+
+	uint32_t ca_cmd_id;
+	uint32_t ca_param_types = 0;
+
+	TEE_Param *ca_params = NULL;
+	size_t ca_params_size;
+	size_t ca_params_expected_size;
+
+	struct mobj *mobj = NULL;
+	size_t mobj_sz = 0;
+	void *mobj_va = NULL;
+
+	uint32_t ca_cmd_ret;
+	uint32_t ca_cmd_ret_origin;
+
+	struct thread_param rpc_params[THREAD_RPC_MAX_NUM_PARAMS];
+	size_t rpc_num_params;
+
+	TEE_Result res;
+
+	/* Check TA interface parameter types */
+	if (pt1 != TEE_PARAM_TYPE_VALUE_INOUT ||
+	    !(pt2 == TEE_PARAM_TYPE_NONE ||
+	      pt2 == TEE_PARAM_TYPE_MEMREF_INOUT) ||
+	    pt3 != TEE_PARAM_TYPE_NONE ||
+	    pt4 != TEE_PARAM_TYPE_NONE) {
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	/* Extract the parameters from the TA interface */
+	ca_cmd_id = params[0].value.a;
+
+	/* Process OCALL parameters, if any */
+	if (pt2 == TEE_PARAM_TYPE_MEMREF_INOUT) {
```

Comment:
How about skipping this optimization and instead always require the TA to always supply this?

Reply:
Section 4.9.4 of the TEE Internal Core API v1.2.1 spec. states about calling `TEE_InvokeTACommand`:

> Unless all parameter types are set to `TEE_PARAM_TYPE_NONE`, `params` SHALL NOT be `NULL` [...]

I'd like `TEE_InvokeCACommand` to behave as closely to `TEE_InvokeTACommand` as possible. So, `params` can be `NULL` if all parameter types are `NONE`. When this is the case, to remove the check you highlighted, `TEE_InvokeCACommand` would have to pass a dummy array to the PTA.

Reply:
Done.

Reply:
`TEE_InvokeCACommand()` passing a dummy array sounds good.